ci/scripts/find-regression.py

```diff
@@ -0,0 +1,262 @@
+        given commit 0->1(start)->2->3(bad)->4, start will be 1, end will be 3. Then the middle commit is (1+3)//2 = 2
+        given commit 0->1(start)->2(bad), start will be 1, end will be 2. Then the middle commit is (1+2)//2 = 1.
+        given commit 0->1(start,bad), start will be 1, end will be 1. We just return the bad commit (1) immediately.
+2. Run the pipeline on the middle commit.
+3. If the pipeline fails, the regression is in the first half of the commits. Recurse (start, mid)
+4. If the pipeline passes, the regression is in the second half of the commits. Recurse (mid+1, end)
+5. If start>=end, return start as the regressed commit.
+
+We won't run the entire pipeline, only steps specified by the BISECT_STEPS environment variable.
+
+For step (2), we need to check its outcome and only run the next step, if the outcome is successful.
+'''
+
+
+def format_step(env):
+    commit = get_bisect_commit(env["START_COMMIT"], env["END_COMMIT"])
+    step = f'''
+cat <<- YAML | buildkite-agent pipeline upload
+steps:
+  - label: "run-{commit}"
+    key: "run-{commit}"
+    trigger: "main-cron"
+    soft_fail: true
+    build:
+      branch: {env["BISECT_BRANCH"]}
+      commit: {commit}
+      env:
+        CI_STEPS: {env['BISECT_STEPS']}
+  - wait
+  - label: 'check'
+    command: |
+        START_COMMIT={env['START_COMMIT']} END_COMMIT={env['END_COMMIT']} BISECT_BRANCH={env['BISECT_BRANCH']} BISECT_STEPS=\'{env['BISECT_STEPS']}\' ci/scripts/find-regression.py check
+YAML'''
+    return step
+
+
+def report_step(commit):
+    step = f'''
+cat <<- YAML | buildkite-agent pipeline upload
+steps:
+  - label: "Regressed Commit: {commit}"
+    command: "echo 'Regressed Commit: {commit}'"
+YAML'''
+    print(f"--- reporting regression commit: {commit}")
+    result = subprocess.run(step, shell=True)
+    if result.returncode != 0:
+        print(f"stderr: {result.stderr}")
+        print(f"stdout: {result.stdout}")
+        sys.exit(1)
+
+
+# Triggers a buildkite job to run the pipeline on the given commit, with the specified tests.
+def run_pipeline(env):
+    step = format_step(env)
+    print(f"--- running upload pipeline for step\n{step}")
+    result = subprocess.run(step, shell=True)
+    if result.returncode != 0:
+        print(f"stderr: {result.stderr}")
+        print(f"stdout: {result.stdout}")
+        sys.exit(1)
+
+
+# Number of commits for [start, end)
+def get_number_of_commits(start, end):
+    cmd = f"git rev-list --count {start}..{end}"
+    result = subprocess.run([cmd], shell=True, capture_output=True, text=True)
+    if result.returncode != 0:
+        print(f"stderr: {result.stderr}")
+        print(f"stdout: {result.stdout}")
+        sys.exit(1)
+    return int(result.stdout)
+
+
+def get_bisect_commit(start, end):
+    number_of_commits = get_number_of_commits(start, end)
+    commit_offset = number_of_commits // 2
+    if commit_offset == 0:
+        return start
```

Comment:
Why not using `git bisect` directly, but manually calculate here?

Reply:
Because `git bisect` means the entire bisect logic has to run within the script itself.

This means we need to take care of synchronizing with buildkite. E.g. we have to keep polling `buildkite step get`, to check if the step is done or not.

I want to just reuse the synchronization logic of buildkite, via the `wait` step. Hence I embed the logic instead.

Reply:
Oh, I finally got how it works after seeing the graph. 🤣

<img width="522" alt="image" src="https://github.com/user-attachments/assets/8a9f39da-e018-472a-8a1c-fac95e08b1a1">

Reply:
But I still don't fully understand why "synchronizing with buildkite" is needed? Why can't we just do `git bisect` in one step? Like how you will `git bisect` locally.

Reply:
Perhaps it's acceptable to poll every 30s for a step instead. Let me try it. Potentially it can simplify the script.

Reply:
> But I still don't fully understand why "synchronizing with buildkite" is needed? Why can't we just do git bisect in one step? Like how you will git bisect locally.

Synchronizing with buildkite is needed, because we always need to wait for the outcome of a step before running the next one. So checking with buildkite whether a step is done is the "synchronization". There's two ways that can be done:
1. step get.
2. Use `wait`, but then the bisect logic must be embedded.

>  Like how you will git bisect locally.

I suppose it's possible. `git bisect` is originally stateful in normal use. Because each step does not share state the typical use of git bisect doesn't work.

I guess we can reinitialize the good and bad commit per step, to make it stateless. Let me have a try for that. Then we don't have to deal with polling buildkite.

Reply:
Well, my question is still: why do we need multiple Buildkite `step`s...?

Reply:
Oh, because one bisect "step" is triggering the pipeline. I finally get it 😅

Reply:
![Screenshot 2024-07-17 at 1 00 19 PM](https://github.com/user-attachments/assets/5ebc3380-42d9-49fe-8622-ce4aa1cd7f69)

I tried the polling approach. Seems like there's a deadlock. The triggered step refuses to run because the bisect step which triggers all the builds is not finished. See the screenshot above. I suppose it's a quirk of buildkite. I tried setting `async: true` as well. No difference in behaviour. Since we rely on bisect step being a long running task, and triggered steps to complete before it, this approach doesn't work.

Using `git bisect` per step is also still hacky. Because we need to manually parse the output from git bisect, handle error states of git bisect, find commit bisecting.

I will stick to the current approach.